### PR TITLE
Closes #2967, Closes #1138, Closes #3025: Upgrade Drupal core to 10.2.x.

### LINKED ISSUE #1138: Problems with Files/Media in Drupal 9
## Problems
There a number of issues with files/media in Drupal 9, but some of them pose issues for a lot of larger sites, or sites with legacy content that will be migrating to Quickstart 2.0. Using this ticket to document these problems and hopefully we find solutions to them over time.

### 1.) Files persisting after deletion

Currently in Quickstart 2, the only way to replace a file is to delete that file and then upload a new file with the same file name. However, that method doesn't work properly due to files persisting after deletion. If you navigate to `/admin/content/media`, and you delete a document/file, that file still persists in the "Files" tab (`/admin/content/files`) and can still be accessed via URL and via the files directory. If you upload another document with the same file name (after deleting the first one) the file name will still be automatically incremented to use _0, _1, _2, etc. These files persist even after numerous cache clears and cron runs. And the Files tab doesn't allow you to actually manage your files (delete, edit, etc).

#### Related Drupal Issues

- https://www.drupal.org/project/drupal/issues/2936175
- https://www.drupal.org/project/drupal/issues/3027324

----------------------------------------

### 2.) No way to replace files

Not being able to replace files will be a problem for a lot of sites that frequently update documents (such as the Registrar's Office). This is related to the above issue as well as the new path naming convention that Drupal 9 uses, which includes the upload date's year and month as the containing directory for that file. With deleting a file being the only way to replace a file, this means the link to view that file will change if the document is uploaded in a different month than the original version of that file.

#### Possible Solutions

- [Media Entity File Replace contrib module](https://www.drupal.org/project/media_entity_file_replace), which allows you to actually replace an existing file, and also allows you to elect whether you would like to preserve the existing file path and file name.

----------------------------------------

### 3.) Can't access revisions for media entities

Media now comes with revisions, but there is currently no way to access those revisions.

#### Related Drupal Issues

- https://www.drupal.org/project/drupal/issues/2911977 (closed as a duplicate of 2350939)
- https://www.drupal.org/project/drupal/issues/2350939

### LINKED ISSUE #2967: Upgrade Drupal core to 10.2.x
## Motivation
Drupal 10.2.0 will be released Dec 11 (see https://www.drupal.org/about/core/policies/core-release-cycles/schedule#current).

## Proposed Resolution
We should update Quickstart to use 10.2.0 ASAP after that release.

## Additional context
https://www.drupal.org/project/drupal/releases/10.2.0-alpha1
https://www.drupal.org/project/drupal/releases/10.2.0-beta1
https://www.drupal.org/project/drupal/releases/10.2.0-rc1

### LINKED ISSUE #3025: Drupal 10.2.0-rc1 Field Widget Core Delete Breaks Cards
## Problem/Motivation
A long-awaited core feature, Drupal 10.2 seems to introduce [remove buttons on multiple value fields.](https://www.drupal.org/project/drupal/issues/1038316)

Presently, our custom card field widget institutes a workaround for this feature not having been available previously, in the form of a custom removal button. The implementation of our custom callback and the core callback are similar in approach, but conflict with each other.

### To Reproduce
On #2968 , create a flexible page with a card deck and attempt to delete an individual card with either removal button. An ajax error results.

## Proposed resolution
Remove our custom remove button and fall back on core behavior.

### Expected behavior
Core removal button should remove the card.

### PR DESCRIPTION
**Pull request by joeparsons**
## Description

Upgrades Drupal core to 10.2.x (this will be a 2.9.x minor release only change).

Companion scaffolding repo PR: https://github.com/az-digital/az-quickstart-scaffolding/pull/135
Companion dev metapackage PR: https://github.com/az-digital/az-quickstart-dev/pull/102

### Changes:
- Updates [core SVG logo support patch](https://drupal.org/i/2259567)
- Updates media ~and block~ permissions (see change record: [Media now provides a UI for viewing, reverting and deleting revisions](https://www.drupal.org/node/3366630))
- Adds `showBlocks` plugin to editor configs (see change record: [CKEditor 5 now offers "Show Blocks" functionality](https://www.drupal.org/node/3388811))
- Adds `fetchpriority=high` to site logos (see change record: [Core themes add fetchpriority=high to the site logo, to improve page rendering performance](https://www.drupal.org/node/3373012))
- Applies config differences from re-exporting config using @tadean's [drush command](https://github.com/az-digital/az_quickstart/pull/2540)
  - Related issue: [Give display modes a description](https://www.drupal.org/project/drupal/issues/3364659)

#### 10.2 Compatibility / Deprecation fixes
- Replaces `file_validate_*` functions with `file.validator` plugins (see change record: [file_validate and related functions are deprecated and replaced with file.validator service and Constraint plugins](https://www.drupal.org/node/3363700))
- Adds new constructor argument to forms extending `ConfigFormBase` (see change record: [New parameter added to \Drupal\Core\Form\ConfigFormBase::__construct](https://www.drupal.org/node/3404140))
- Replaces block plugin annotations with block attributes (see change record: [Plugin implementations should use PHP attributes instead of annotations](https://www.drupal.org/node/3395575))

Also includes changes from #3026 to make the card field widget work with upstream field UI changes from Drupal core.

#### Enhancements based on Drupal core changes that will be handled separately
- #3015
- #3016
- #3018
- #3019
- #3011

## Related issues

Closes #2967
Closes #1138 
Closes #3025

## How to test

Link to Probo review site: https://7b5e2985-cedf-4a13-ab91-3a35916c835d--pr-2968.probo.build/

- General testing
- Ensure uploading an SVG logo still works in the theme settings
- Ensure Quickstart settings form and AZ News Feeds admin forms still work
- Ensure author revision pages still work
- Ensure file validation still works for Bibtex import form and footer logo image in theme settings form
- Ensure the Quickstart Carousel Notification block and  Quickstart select menu block still work
- Ensure that `fetchpriority=high` is added to the institutional wordmark and site logo image elements
- Ensure media ~and block~ revisions are accessible for content editors/admins
- Ensure "show blocks" button shows up in ckeditors

### To be tested with separate PRs
- ~Ensure logging still works for config override imports~
- ~Ensure publications import menu page still works~
- ~Ensure Trellis Events import form still works~
- ~Ensure AZ Metrics logging still works~
- ~Ensure AZ News feed importer still works~
- ~Ensure AZ Publications importer still works~
- ~Ensure AZ Core monitoring page route still works~

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [ ] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [x] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -160,7 +160,7 @@
             },
             "drupal/core": {
                 "Views Date Filter Datetime Granularity Option (2868014)": "https://www.drupal.org/files/issues/2022-11-16/2868014-127.datetime-granularity.patch",
-                "Unnecessary restrictions on logo format: Can't upload replacement SVG logo (2259567)": "https://www.drupal.org/files/issues/2022-05-05/2259567-160-9-3-x.patch",
+                "Unnecessary restrictions on logo format: Can't upload replacement SVG logo (2259567)": "https://www.drupal.org/files/issues/2023-05-22/drupal-n2259567-173.patch",
```
Comment (joeparsons) [inline composer.json:163]:
```suggestion
                "Support SVG files for theme logo setting (2259567)": "https://www.drupal.org/files/issues/2023-05-22/drupal-n2259567-173.patch",
```

**Comment by trackleft** [comment]
Do you want to update our codebase: https://github.com/search?q=repo%3Aaz-digital%2Faz_quickstart%20event_subscriber&type=code


See: https://www.drupal.org/node/3357408

**Comment by trackleft** [comment]
Do we want to add a file.settings.yml with some transliteration settings configured?
https://www.drupal.org/node/2972665

**Comment by trackleft** [comment]
Should we update our code to take https://www.drupal.org/node/3395716 into account?
See https://github.com/search?q=repo%3Aaz-digital%2Faz_quickstart%20public%20static%20function%20create&type=code

**Comment by trackleft** [comment]
Looks like we need to update permissions

Users with the 'administer media' permission will have access without needing to grant the new permissions.
User with 'access media overview' and 'view any $bundle media history' permission can view revision history.
User with 'access media overview' and 'revert any $bundle media revisions' permission can revert to that revision.
User with 'access media overview' and 'delete any $bundle media revisions' permission can delete that revision.

See: https://www.drupal.org/node/3366630

File: modules/custom/az_core/az_core.services.yml [hunk 2]
```diff
@@ -9,4 +12,4 @@ services:
     arguments: ['az_core']
   az_core.override_import:
     class: Drupal\az_core\AZConfigOverride
-    arguments: ['@config.factory', '@extension.list.module', '@config_provider.collector', '@config_sync.snapshotter', '@config_update.config_list', '@module_handler', '@logger.channel.az_core']
+    arguments: ['@config.factory', '@extension.list.module', '@config_provider.collector', '@config_sync.snapshotter', '@config_update.config_list', '@module_handler']
```
Comment (trackleft) [inline modules/custom/az_core/az_core.services.yml:15]: I wish we did this:

```suggestion
    arguments: 
    - '@config.factory', 
    - '@extension.list.module',
    - '@config_provider.collector',
    - '@config_sync.snapshotter',
    - '@config_update.config_list',
    - '@module_handler'
```

**Comment by trackleft** [comment]
Core themes add `fetchpriority=high` to the site logo img element, to improve page rendering performance as of https://www.drupal.org/node/3373012

We can probably follow suite by adding the attribute here:
https://github.com/az-digital/az_quickstart/blob/7ecb796ae99e3f38bd1005f5e8bbef84ab90fa12/themes/custom/az_barrio/az_barrio.theme#L441-L448

File: modules/custom/az_publication/src/Controller/AZAuthorController.php [hunk 3]
```diff
@@ -5,8 +5,9 @@
 use Drupal\az_publication\Entity\AZAuthorInterface;
 use Drupal\Component\Utility\Xss;
 use Drupal\Core\Controller\ControllerBase;
-use Drupal\Core\DependencyInjection\ContainerInjectionInterface;
+use Drupal\Core\Datetime\DateFormatterInterface;
 use Drupal\Core\Link;
+use Drupal\Core\Render\RendererInterface;
 use Drupal\Core\Url;
 use Symfony\Component\DependencyInjection\ContainerInterface;
```
Comment (trackleft) [inline modules/custom/az_publication/src/Controller/AZAuthorController.php]:
```suggestion
```

Comment (trackleft) [reply modules/custom/az_publication/src/Controller/AZAuthorController.php]: Oops I left that in there.

File: modules/custom/az_core/az_core.services.yml [hunk 4]
```diff
@@ -1,4 +1,7 @@
 services:
+  _defaults:
+    autoconfigure: true
+
```
Comment (trackleft) [inline modules/custom/az_core/az_core.services.yml:4]:
```suggestion
```

**Comment by trackleft** [comment]
Test that cards can be deleted from card decks.
See https://github.com/az-digital/az_quickstart/issues/3010
See https://www.drupal.org/project/drupal/issues/1038316

**Comment by trackleft** [comment]
Interestingly, we can probably remove our delete button 
image

**Comment by trackleft** [comment]
The "remove" button was introduced by this pull request.

**Comment by danahertzberg** [comment]
General testing - in progress

- [ ] Ensure uploading an SVG logo still works in the theme settings
- [ ] Ensure logging still works for config override imports
- [ ] Ensure Quickstart settings form and AZ News Feeds admin forms still work
- [ ] Ensure publications import menu page still works
- [ ] Ensure author revision pages still work
- [ ] Ensure file validation still works for Bibtex import form and footer logo image in theme settings form
- [ ] Ensure the Quickstart Carousel Notification block and Quickstart select menu block still work
- [ ] Ensure that fetchpriority=high is added to the institutional wordmark and site logo image elements
- [ ] Ensure Trellis Events import form still works
- [ ] Ensure AZ Metrics logging still works
- [ ] Ensure AZ News feed importer still works
- [ ] Ensure AZ Publications importer still works
- [ ] Ensure media and block revisions are accessible for content editors/admins
- [ ] Ensure AZ Core monitoring page route still works
- [ ] Ensure "show blocks" button shows up in ckeditors

**Comment by bberndt-uaz** [comment]
## Testing (In Progress)

- [ ] General testing
- [x] Ensure uploading an SVG logo still works in the theme settings
  - Uploading an SVG logo works. As @danahertzberg noted, the Inline logo SVG option does not work. Trying to access a page afterwards results in a white screen with this error: `TypeError: array_intersect_key(): Argument #1 ($array) must be of type array, string given in array_intersect_key() (line 348 of core/lib/Drupal/Core/Render/Renderer.php)`.
- [ ] Ensure logging still works for config override imports
- [x] Ensure Quickstart settings form and AZ News Feeds admin forms still work
- [ ] Ensure publications import menu page still works
- [ ] Ensure author revision pages still work
- [ ] Ensure file validation still works for Bibtex import form and footer logo image in theme settings form
- [x] Ensure the Quickstart Carousel Notification block and Quickstart select menu block still work
- [x] Ensure that fetchpriority=high is added to the institutional wordmark and site logo image elements
- [ ] Ensure Trellis Events import form still works
- [ ] Ensure AZ Metrics logging still works
- [x] Ensure AZ News feed importer still works
- [ ] Ensure AZ Publications importer still works
- [ ] Ensure media and block revisions are accessible for content editors/admins
- [ ] Ensure AZ Core monitoring page route still works
- [ ] Ensure "show blocks" button shows up in ckeditors

**Comment by joeparsons** [comment]
The inline SVG problems seem to be an existing issue (#1758)

File: config/install/user.role.az_content_editor.yml [hunk 5]
```diff
@@ -78,18 +78,21 @@ permissions:
   - 'edit any az_carousel_item content'
   - 'edit any az_document media'
   - 'edit any az_event content'
+  - 'edit any az_flexible_block block content'
```
Comment (joeparsons) [inline config/install/user.role.az_content_editor.yml:81]:
```suggestion
```

File: config/install/user.role.az_content_editor.yml [hunk 6]
```diff
@@ -78,18 +78,21 @@ permissions:
...
   - 'edit any az_flexible_page content'
   - 'edit any az_image media'
   - 'edit any az_news content'
   - 'edit any az_person content'
   - 'edit any az_publication content'
+  - 'edit any az_quick_links block content'
```
Comment (joeparsons) [inline config/install/user.role.az_content_editor.yml:87]:
```suggestion
```

**Review by trackleft — APPROVED** [review]

**Comment by danahertzberg** [comment]
## General testing V2 (in progress)

- [ ] Ensure uploading an SVG logo still works in the theme settings
- [ ] Ensure Quickstart settings form and AZ News Feeds admin forms still work
- [ ] Ensure author revision pages still work
- [ ] Ensure file validation still works for Bibtex import form and footer logo image in theme settings form
- [ ] Ensure the Quickstart Carousel Notification block and Quickstart select menu block still work
- [ ] Ensure that fetchpriority=high is added to the institutional wordmark and site logo image elements
- [ ] Ensure media and block revisions are accessible for content editors/admins
- [ ] Ensure "show blocks" button shows up in ckeditors

**Review by trackleft — APPROVED** [review]

File: config/install/user.role.az_content_editor.yml [hunk 7]
```diff
@@ -111,6 +112,13 @@ permissions:
   - 'revert all author revisions'
   - 'revert all revisions'
   - 'revert az_carousel_item revisions'
+  - 'revert any az_audio media revisions'
+  - 'revert any az_document media revisions'
+  - 'revert any az_flexible_block block content revisions'
```
Comment (joeparsons) [inline config/install/user.role.az_content_editor.yml:117]:
```suggestion
```

File: config/install/user.role.az_content_editor.yml [hunk 8]
```diff
@@ -111,6 +112,13 @@ permissions:
...
+  - 'revert any az_image media revisions'
+  - 'revert any az_quick_links block content revisions'
```
Comment (joeparsons) [inline config/install/user.role.az_content_editor.yml:119]:
```suggestion
```

File: config/install/user.role.az_content_editor.yml [hunk 9]
```diff
@@ -133,6 +141,13 @@ permissions:
   - 'view all author revisions'
   - 'view all media revisions'
   - 'view all revisions'
+  - 'view any az_audio media revisions'
+  - 'view any az_document media revisions'
+  - 'view any az_flexible_block block content history'
```
Comment (joeparsons) [inline config/install/user.role.az_content_editor.yml:146]:
```suggestion
```

File: config/install/user.role.az_content_editor.yml [hunk 10]
```diff
@@ -133,6 +141,13 @@ permissions:
...
+  - 'view any az_image media revisions'
+  - 'view any az_quick_links block content history'
```
Comment (joeparsons) [inline config/install/user.role.az_content_editor.yml:148]:
```suggestion
```

**Review by tadean — APPROVED** [review]

**Review by danahertzberg — APPROVED** [review]

**Pull request merged**